src/Core/src/ImageSources/iOS/ImageAnimationHelper.cs

```diff
@@ -0,0 +1,216 @@
+// https://github.com/luberda-molinet/FFImageLoading/blob/bb675c6011b39ddccecbe6125d1853de81e6396a/source/FFImageLoading.Shared.IosMac/Decoders/GifDecoder.cs
+
+using System;
+using ImageIO;
+using Foundation;
+using CoreGraphics;
+using UIKit;
+
+namespace Microsoft.Maui;
+
+static class ImageAnimationHelper
+{
+	sealed class ImageDataHelper : IDisposable
+	{
+		readonly nfloat _scale;
+		readonly CGImage?[] _keyFrames;
+		readonly int[] _delayTimes;
+		readonly int _imageCount;
+		int _totalAnimationTime;
+		bool _disposed;
+
+		public ImageDataHelper(nint imageCount, nfloat scale)
+		{
+			if (imageCount <= 0)
+				throw new ArgumentOutOfRangeException(nameof(imageCount), $"{nameof(imageCount)} is 0, no images to animate.");
+			if (scale < 1)
+				throw new ArgumentOutOfRangeException(nameof(scale), $"{nameof(scale)} is < 1, cannot scale up.");
+
+			_scale = scale;
+			_keyFrames = new CGImage[imageCount];
+			_delayTimes = new int[imageCount];
+			_imageCount = (int)imageCount;
+
+			Width = 0;
+			Height = 0;
+		}
+
+		public int Width { get; private set; }
+
+		public int Height { get; private set; }
+
+		public void AddFrameData(CGImageSource imageSource, int index)
+		{
+			if (index < 0 || index >= _imageCount || index >= imageSource.ImageCount)
+				throw new ArgumentOutOfRangeException(nameof(index), $"Error adding frame data. {nameof(index)} is less than 0, or more than or equal to image count.");
+
+			var imageProperties = imageSource.GetProperties(index, null);
+			using var gifImageProperties = imageProperties?.Dictionary[ImageIO.CGImageProperties.GIFDictionary];
+			using var unclampedDelayTimeValue = gifImageProperties?.ValueForKey(ImageIO.CGImageProperties.GIFUnclampedDelayTime);
+			using var delayTimeValue = gifImageProperties?.ValueForKey(ImageIO.CGImageProperties.GIFDelayTime);
+
+			var delayTime = 0.1;
+			if (unclampedDelayTimeValue is NSNumber unclampedDelay)
+				delayTime = unclampedDelay.DoubleValue;
+			else if (delayTimeValue is NSNumber delay)
+				delayTime = delay.DoubleValue;
+
+			var image = imageSource.CreateImage(index, null!);
+			if (image is null)
+				throw new ArgumentException($"Image source did not contain an image at index {index}.");
+
+			AddFrameData(image, index, delayTime);
+		}
+
+		public void AddFrameData(CGImage image, int index, double delayTime = 0.1)
+		{
+			if (index < 0 || index >= _imageCount)
+				throw new ArgumentOutOfRangeException(nameof(index), $"Error adding frame data. {nameof(index)} is less than 0, or more than or equal to image count.");
+			if (image is null)
+				throw new ArgumentNullException(nameof(image));
+
+			// Frame delay compability adjustment.
+			if (delayTime <= 0.02)
+				delayTime = 0.1;
+
+			// GIF only has centiseconds data
+			var centiseconds = (int)(delayTime * 100.0);
+
+			Width = Math.Max(Width, (int)image.Width);
+			Height = Math.Max(Height, (int)image.Height);
+
+			_keyFrames[index]?.Dispose();
+			_keyFrames[index] = null;
```

Comment:
Do we need to set this to null right here?

If we use `null!` in dispose then I think we can just make "_keyFrames"  `CGImage[]` instead of `CGImage?[]` and then don't need null forgiveness here

![image](https://github.com/dotnet/maui/assets/5375137/8d1b9144-76db-4f6d-83db-b7049178071d)